File: On_Site_test_2.cpp
#include <iostream>
#include <string>
#include <sstream>
#include <fstream>
#include <vector>
#include <map>
#include<stack>
#include<iomanip>
#include<algorithm>
using namespace std;
class SIC_Line
{
public:
	string Address_label;
	string Op_code;
	string Operant;
	SIC_Line(string line)
	{
		stringstream ss;
		ss << line;
		getline(ss, Address_label, '\t');
		getline(ss, Op_code, '\t');
		getline(ss, Operant, '\t');
	}
};
class SIC_MACRO
{
public:
	string unique_Labels = "AA";
	vector<string> ARGTAB;
	vector<string> KEYWORDTAB;//自創
	vector<SIC_Line> Lines;
	void init_Macro()//回傳巨集行數
	{
		vector<string> parameters;//&開頭為參數名稱
		string p;
		stringstream ss; ss << Lines[0].Operant;
		while (getline(ss, p, ','))
		{
			stringstream ss2; ss2 << p;
			string p2;
			getline(ss2, p2, '=');
			KEYWORDTAB.push_back(p2);//儲存參數名稱
			getline(ss2, p2);
			ARGTAB.push_back(p2);//若有等號 儲存另一端的預設參數
		}
		for (int i=0;i<KEYWORDTAB.size();i++)//取代每個&參數為? 未完成->拼接
		{
			string restring = "?" + to_string(i);
			for (int o = 1; o < Lines.size(); o++)
			{	
				int findIndex = Lines[o].Operant.find(KEYWORDTAB[i]);
				//若該欄字串完全吻合 或是部分字串吻合且下一個字元不為字母 E.G X'&EOR'為真 &EORCK為假
				if ((Lines[o].Operant == KEYWORDTAB[i])||findIndex != -1 && ((findIndex + KEYWORDTAB[i].length())==Lines[o].Operant.length()||!isalpha(Lines[o].Operant[findIndex + KEYWORDTAB[i].length()]))) Lines[o].Operant.replace(Lines[o].Operant.find(KEYWORDTAB[i]), KEYWORDTAB[i].length(), restring);
			}
		}
		
	}
	void extend_Macro(vector<SIC_Line>& target,string head_Address_Label, string operant,map<string,SIC_MACRO> macromap)
	{
		vector<string> args;
		map<string, string> IF_keyword;
		stringstream ss;
		string s;
		ss << operant;
		vector<string> build_Parameters;//處理參數 keyword/無keyword兩種
		while (getline(ss, s, ','))build_Parameters.push_back(s);//使用,拆分字串
		for (auto& L : ARGTAB)args.push_back(L);//複製預設參數
		if (build_Parameters[0].find("=") == -1)//如果沒有=則為依序填入
		{
			for (int i = 0; i < build_Parameters.size(); i++)args[i] = build_Parameters[i];
		}
		else//否則依照關鍵字填入
		{
			for (auto& p : build_Parameters)
			{
				stringstream ss; ss << p;
				string keyword;
				getline(ss, keyword, '=');
				keyword = "&" + keyword;
				auto loc= find(KEYWORDTAB.begin(), KEYWORDTAB.end(), keyword);
				int index = loc - KEYWORDTAB.begin();
				getline(ss, args[index]);
			}
		}
		ofstream ARGTAB("ARGTAB.txt", ios::out);//輸出ARGTAB 沒錯 在這裡 他叫用一次巨集輸出一個 除非他說要生成多個ARGTAB不然我都讓他被覆蓋掉
		for (auto arg : args)ARGTAB << arg << endl;
		ARGTAB.close();

			//TODO IF判斷式
		Lines[1].Address_label = head_Address_Label;//替換首行的address_label
		bool if_skipping = false;//當此flag被標記為true跳過輸出直到else或endif
		bool else_outputting = false;//當此flag標記為true則輸出直到endif
		stack<string> IF_statment;

		bool first = true;//紀錄是否為首行(避免IF判斷式)
		for (int Lcount=1; Lcount < Lines.size()-1; Lcount++)
		{
			SIC_Line L(Lines[Lcount]);
			if (L.Op_code == "SET")continue;
			if (L.Op_code == "IF")continue;
			if (L.Op_code == "ELSE")continue;
			if (L.Op_code == "ENDIF")continue;
			//取代? 最大支援到?9 我是沒心力寫下去了
			if (L.Operant.find("?") != -1)
			{
				int replaceIndex = L.Operant.find("?");
				int keywordIndex = L.Operant[replaceIndex + 1] - '0';
				L.Operant.replace(replaceIndex, 2, args[keywordIndex]);
			}
			//流水號
			if (L.Address_label[0] == '$')
				L.Address_label = "$" + unique_Labels + L.Address_label.substr(1, L.Address_label.length() - 1);
			if (L.Operant[0] == '$')L.Operant = "$" + unique_Labels + L.Operant.substr(1, L.Operant.length() - 1);

			if (first)
			{
				L.Address_label = head_Address_Label;
				first = false;
			}
			if (macromap.find(L.Op_code) != macromap.end())
			{
				L.Address_label = "." + L.Address_label;
				target.push_back(L);
				macromap[L.Op_code].extend_Macro(target, L.Address_label, L.Operant, macromap);
			}
			else target.push_back(L);
		}
		//TOFO:SET IF ELSE WHILE ARRAY
		next_Unique_Labels();//切換至下一個流水號
	}
	void next_Unique_Labels()
	{
		unique_Labels[1]++;
		if (unique_Labels[1] > 'Z')
		{
			unique_Labels[0]++;
			unique_Labels[1] = 'A';
		}
	}
};

class SIC_Program
{
public:
	vector<SIC_Line> Lines;
	vector<SIC_Line> output_Lines;
	map<string, SIC_MACRO> Macros;
	void addline(string line)
	{
		Lines.push_back(SIC_Line(line));
	}
	void print()
	{
		ofstream output("output.txt", ios::out);
		for (auto L : output_Lines)output << L.Address_label << "\t" << L.Op_code << "\t" << L.Operant << endl;
		output.close();
	}
	void pass()
	{
		ofstream DEFTAB("DEFTAB.txt", ios::out);
		ofstream NAMTAB("NAMTAB.txt", ios::out);
		int DEFLine = 0;
		stack<string> target_Macro;
		for (auto& Line : Lines)
		{
			if (Line.Op_code == "MACRO")
			{
				target_Macro.push(Line.Address_label);//將目標MACRO放入堆疊
				vector<string> parameters;
				Macros[Line.Address_label] = SIC_MACRO();
				Macros[target_Macro.top()].Lines.push_back(Line);//輸入首行
			}
			else if (Line.Op_code == "MEND")
			{
				Macros[target_Macro.top()].Lines.push_back(Line);//輸入尾行
				Macros[target_Macro.top()].init_Macro();
				NAMTAB << target_Macro.top() << "," << DEFLine << ",";
				DEFLine += Macros[target_Macro.top()].Lines.size();
				for (auto L : Macros[target_Macro.top()].Lines)DEFTAB << L.Op_code << "\t" << L.Operant << endl;
				NAMTAB << DEFLine - 2 << endl;
				target_Macro.pop();
			}
			else if (Macros.find(Line.Op_code) != Macros.end() && target_Macro.empty())//若不在巨集建立模式(堆疊為空)則展開巨集
			{
				output_Lines.push_back(Line);
				output_Lines.back().Address_label = "." + output_Lines.back().Address_label;
				Macros[Line.Op_code].extend_Macro(output_Lines, Line.Address_label, Line.Operant,Macros);
			}
			else
			{
				if (target_Macro.empty())output_Lines.push_back(Line);
				else Macros[target_Macro.top()].Lines.push_back(Line);
			}
		}
		DEFTAB.close();
		NAMTAB.close();
	}
};

int main()
{
	ifstream input_file("input01.txt", ios::in); 
	string s;
	SIC_Program sic_Program;
	while (getline(input_file,s))
	{
		sic_Program.addline(s);
	}
	input_file.close();
	sic_Program.pass();
	sic_Program.print();
	return 0;
}

File: On_Site_test_2.py
from copy import deepcopy
import sys

class SIC_Line:

    address_Label = str()
    Op_code = str()
    Operant = str()

    def __init__(self, line: str):

        temp_cmd = line.removesuffix('\n').split('\t')
        length = len(temp_cmd)
        if len(temp_cmd) == 1:
            self.address_Label = temp_cmd[0]
            self.Op_code = ""
            self.Operant = ""

        elif len(temp_cmd) == 2:
            self.address_Label = temp_cmd[0]
            self.Op_code = temp_cmd[1]
            self.Operant = ""
        else:
            self.address_Label = temp_cmd[0]
            self.Op_code = temp_cmd[1]
            self.Operant = temp_cmd[2]


class SIC_MACRO:

    unique_Labels = "AA"
    KEYWORDTAB = list()
    ARGTAB = list()
    Lines = list()

    def init_Macro(self):

        p = str()
        ss = self.Lines[0].Operant
        ss2 = ss.split(',')
        for s in ss2:
            self.KEYWORDTAB.append(s.split('=')[0])  # 儲存參數名稱
            self.ARGTAB.append(s.split('=')[1] if len(
                s.split('=')) > 1 else "")  # 若有等號 儲存另一端的預設參數

        for i in range(len(self.KEYWORDTAB)):  # 取代每個&參數為? 未完成->拼接
            restring = "?" + str(i)
            for o in range(1, len(self.Lines)):
                findIndex = self.Lines[o].Operant.find(self.KEYWORDTAB[i])
                if findIndex != -1:
                    print("findIndex")
                # 若該欄字串完全吻合 或是部分字串吻合且下一個字元不為字母 E.G X'&EOR'為真 &EORCK為假
                if self.Lines[o].Operant == self.KEYWORDTAB[i] or (findIndex != -1 and ((findIndex + len(self.KEYWORDTAB[i])) == len(self.Lines[o].Operant) or (not self.Lines[o].Operant[findIndex + len(self.KEYWORDTAB[i])].isalpha()))):
                    self.Lines[o].Operant = self.Lines[o].Operant.replace(
                        self.KEYWORDTAB[i], restring)

    def extend_Macro(self, target, head_Address_Label: str, Operant: str, macromap: dict):
        args = list()
        IF_keyword = dict()
        # 處理參數 keyword/無keyword兩種並使用','拆分字串
        build_Parameters = Operant.split(',')
        args = [L for L in self.ARGTAB]  # 複製預設參數
        if build_Parameters[0].find('=') == -1:  # 如果沒有=則為依序填入
            for i in range(len(build_Parameters)):
                args[i] = build_Parameters[i]
        else:  # 否則依照關鍵字填入
            for p in build_Parameters:
                keyword = "&" + p.split('=')[0]
                index = self.KEYWORDTAB.index(keyword)
                args[index] = p.split('=')[1]
        for arg in args:
            self.ARGTAB.append(arg)

            # TODO IF判斷式
        self.Lines[1].address_Label = head_Address_Label  # 替換首行的address_label
        if_skipping = False  # 當此flag被標記為true跳過輸出直到else或endif
        else_outputting = False  # 當此flag標記為true則輸出直到endif
        IF_statment = list()

        first = True
        for i in range(1, len(self.Lines)-1):
            L = deepcopy(self.Lines[i])
            if L.Op_code in ["SET", "IF", "ELSE", "ENDIF"]:
                continue  # 取代? 最大支援到?9 我是沒心力寫下去了

            if L.Operant.find('?') != -1:
                replaceIndex = L.Operant.find('?')
                keywordIndex = int(L.Operant[replaceIndex + 1])
                # replace the index keywordIndex to the value of args[keywordIndex]
                L.Operant = L.Operant.replace(
                    "?"+str(keywordIndex), args[keywordIndex])
            # 流水號
            if L.address_Label.startswith('$'):
                L.address_Label = L.address_Label.replace(
                    '$', "$" + self.unique_Labels)
            if L.Operant.startswith('$'):
                L.Operant = L.Operant.replace('$', "$" + self.unique_Labels)

            if first:
                L.address_Label = head_Address_Label
                first = False
            if macromap.get(L.Op_code) != None:
                L.address_Label = "." + L.address_Label
                target.append(L)
                macromap[L.Op_code].extend_Macro(
                    target, L.address_Label, L.Operant, macromap)
            else:
                target.append(L)

            # TODO:SET IF ELSE WHILE ARRAY

    def next_Unique_Labels(self):
        # convert self.unique_Labels[0] to char and +1 then convert back to string
        self.unique_Labels = self.unique_Labels[0] + \
            chr(ord(self.unique_Labels[0]) + 1)
        if self.unique_Labels[1] > 'Z':
            self.unique_Labels = chr(ord(self.unique_Labels[0]) + 1) + "A"


class SIC_Program:
    Lines = list()
    output_Lines = list()
    Macros = dict()

    def addline(self, line: SIC_Line):
        self.Lines.append(line)

    def pass_(self):
        DEFTAB = list()
        NAMTAB = list()
        DEFLine = 0
        target_Macro = list()
        for Line in self.Lines:
            if Line.Op_code == "MACRO":
                target_Macro.append(Line.address_Label)  # 將目標MACRO放入堆疊
                self.Macros[Line.address_Label] = SIC_MACRO()
                self.Macros[target_Macro[-1]].Lines.append(Line)  # 輸入首行
            elif Line.Op_code == "MEND":
                self.Macros[target_Macro[-1]].Lines.append(Line)  # 輸入尾行
                self.Macros[target_Macro[-1]].init_Macro()
                NAMTAB.append(target_Macro[-1] + "\t" + str(DEFLine) + ",")
                DEFLine += len(self.Macros[target_Macro[-1]].Lines)
                for L in self.Macros[target_Macro[-1]].Lines:
                    DEFTAB.append(L.Op_code + "\t" + L.Operant + "\n")
                NAMTAB.append(str(DEFLine - 2) + "\n")
                target_Macro.pop()
            # 若不在巨集建立模式(堆疊為空)則展開巨集
            elif self.Macros.get(Line.Op_code) != None and len(target_Macro) == 0:
                self.output_Lines.append(deepcopy(Line))
                self.output_Lines[-1].address_Label = "." + \
                    self.output_Lines[-1].address_Label
                self.Macros[Line.Op_code].extend_Macro(
                    self.output_Lines, Line.address_Label, Line.Operant, self.Macros)
            else:
                if len(target_Macro) == 0:
                    self.output_Lines.append(Line)
                else:
                    self.Macros[target_Macro[-1]].Lines.append(Line)
        with open("DEFTAB.txt", "w") as f:
            for L in DEFTAB:
                f.write(L)
        with open("NAMTAB.txt", "w") as f:
            for L in NAMTAB:
                f.write(L)

    def print(self):
        with open("output.txt", "w") as f:
            for L in self.output_Lines:

                f.write(L.address_Label + "\t" +
                        L.Op_code + "\t" + L.Operant + "\n")


if __name__ == "__main__":
    
    # check python version must be 3.9 or higher
    if (sys.version_info.major != 3 or sys.version_info.minor < 9):
        exit("Python version must be 3.9 or higher")
    
    sic_progrem = SIC_Program()
    with open("input01.txt", "r") as f:
        for line in f:
            if len(line) > 0:  # 避免空行
                sic_progrem.addline(SIC_Line(line))
    sic_progrem.pass_()
    sic_progrem.print()
